clustermesh-apiserver/clustermesh/synchronizer.go

```diff
@@ -66,45 +84,131 @@ func RegisterSynchronizer[T runtime.Object](in syncParams[T]) {
+							store.Synced(ctx, synced)
+							continue
+						}
+						// Filter the event based on namespace global status.
+						// Only required for certain resource types.
+						// Ignore delete events as they should always be processed.
+						if event.Kind != resource.Delete && in.Options.Namespaced {
+							ns := in.Namespacer.ExtractNamespace(event)
+							if ns == "" {
+								logger.Error("Failed to determine namespace for resource event, skipping",
+									logfields.Name, event.Key.Name,
+								)
+								// No way to process this event, just mark done and continue.
+								event.Done(nil)
+								continue
+							}
+							isGlobal, err := in.NamespaceManager.IsGlobalNamespaceByName(ns)
+							if err != nil {
+								logger.Warn("Failed to determine if namespace is global",
+									logfields.Error, err,
+									logfields.K8sNamespace, ns,
+								)
+								// Retry this as it might succeed later.
+								event.Done(err)
+								continue
+							}
+							if !isGlobal {
+								logger.Debug("Skipping resource event as it is not in a global namespace",
+									logfields.Name, event.Key.Name,
+									logfields.K8sNamespace, ns,
+								)
+								// Handle resources transitioning out of global namespaces.
+								// If a resource was previously in a global namespace and is now
+								// in a non-global namespace (e.g.,mutable fields like in CiliumEndpointSlice),
+								// we need to delete it from kvstore. Convert the event to a delete to ensure cleanup.
+								// event.Done will be called later during normal processing.
+								event.Kind = resource.Delete
+							}
+						}
+
+						// No possible errors past this point.
+						event.Done(nil)
+
+						upserts, deletes := in.Converter.Convert(event)
+						for upsert := range upserts {
+							process("resource-event", "upsert", upsert.GetKeyName(), func() error { return store.UpsertKey(ctx, upsert) })
+						}
+						for delete := range deletes {
+							process("resource-event", "delete", delete.GetKeyName(), func() error { return store.DeleteKey(ctx, delete) })
+						}
+					case event, ok := <-namespaceEvents:
+						if !ok {
+							logger.Info("Namespace event channel closed, ignoring future namespace events")
+							namespaceEvents = nil
+							continue
+						}
+						event.Done(nil)
+						for resEvent := range namespaceHandler(in, resourceStore, event) {
+							upserts, deletes := in.Converter.Convert(resEvent)
+							for upsert := range upserts {
+								process("namespace-event", "upsert", upsert.GetKeyName(), func() error { return store.UpsertKey(ctx, upsert) })
+							}
+							for delete := range deletes {
+								process("namespace-event", "delete", delete.GetKeyName(), func() error { return store.DeleteKey(ctx, delete) })
+							}
+						}
+					default:
+						if resourceEvents == nil && namespaceEvents == nil {
+							logger.Info("All event channels closed, stopping synchronization")
+							return nil
+						}
 					}
```

Comment:
This is wrong, because a default select case in a for loop makes it continuously spin. The correct form is

```go
				for resourceEvents != nil || namespaceEvents != nil {
					select {
					...
```

Reply:
Thanks for pointing this out. Fixed.